docs/architecture/adr-019-protobuf-state-encoding.md

```diff
@@ -110,162 +111,155 @@ type Codec interface {
+After [extension discussion](https://github.com/cosmos/cosmos-sdk/issues/6030),
+this was chosen as the preferred alternative to application-level `oneof`s
+as in our original protobuf design. The arguments in favor of `Any` can be
+summarized as follows:
+* `Any` provides a simpler, more consistent client UX for dealing with
+interfaces than app-level `oneof`s that will need to be coordinated more
+carefully across applications. Creating a generic transaction
+signing library using `oneof`s may be cumbersome and critical logic may need
+to be reimplemented for each chain
+* `Any` provides more resistance against human error than `oneof`
+* `Any` is generally simpler to implement for both modules and apps
+
+The main counter-argument to using `Any` centers around its additional space
+and possibly performance overhead. The space overhead could be dealt with using
+compression at the persistence layer in the future and the performance impact
+is likely to be small. Thus, not using `Any` is seem as a pre-mature optimization,
+with user experience as the higher order concern.
+
+Note, that given the SDK's decision to adopt the `Codec` interfaces described
+above, app's can still choose to use `oneof` to encode state and transactions
+but it is not the recommended approach.
+
+### Safe usage of `Any`
+
+By default, the [gogo protobuf implementation of `Any`](https://godoc.org/github.com/gogo/protobuf/types)
+uses global type registration to decode values packed in `Any` into concrete
+go types. This introduces a vulnerability where any malicious module
+in the dependency tree could registry a type with the global protobuf registry
+and cause it to be loaded and unmarshaled by a transaction that referenced
+it in the `type_url` field.
+
+To prevent this, we introduce a type registration mechanism for decoding `Any`
+values into concrete types through the `InterfaceContext` interface which
+bears some similarity to type registration with Amino:
 
-Example:
-
-```protobuf
-// app/codec/codec.proto
-
-import "third_party/proto/cosmos-proto/cosmos.proto";
-import "x/auth/types/types.proto";
-import "x/auth/vesting/types/types.proto";
-import "x/supply/types/types.proto";
-
-message Account {
-  option (cosmos_proto.interface_type) = "*github.com/cosmos/cosmos-sdk/x/auth/exported.Account";
-
-  // sum defines a list of all acceptable concrete Account implementations.
-  oneof sum {
-    cosmos_sdk.x.auth.v1.BaseAccount                      base_account               = 1;
-    cosmos_sdk.x.auth.vesting.v1.ContinuousVestingAccount continuous_vesting_account = 2;
-    cosmos_sdk.x.auth.vesting.v1.DelayedVestingAccount    delayed_vesting_account    = 3;
-    cosmos_sdk.x.auth.vesting.v1.PeriodicVestingAccount   periodic_vesting_account   = 4;
-    cosmos_sdk.x.supply.v1.ModuleAccount                  module_account             = 5;
-  }
-
-  // ...
+```go
+type InterfaceContext interface {
```

Comment:
Why is this called a context? `InterfaceRegistry`?

Reply:
I don't know. I just needed a name. `InterfaceRegistry` sounds good.